gc/base/standard/Scavenger.cpp

```diff
@@ -1463,7 +1463,13 @@ MM_Scavenger::copy(MM_EnvironmentStandard *env, MM_ForwardedHeader* forwardedHea
 	/* Concurrent Scavenger can update forwarding pointer only after the object has been copied
 	 * (since mutator may access the object as soon as forwarding pointer is installed) */
 	}
-	if (originalDestinationObjectPtr == (allowDuplicate ? (destinationObjectPtr = forwardedHeader->setForwardedObject(destinationObjectPtr)) : destinationObjectPtr)) {
+	if (allowDuplicate) {
+		/* On weak memory model, ensure that this candidate copy is visible
+		 * before (potentially) winning forwarding */
+		MM_AtomicOperations::storeSync();
+		destinationObjectPtr = forwardedHeader->setForwardedObject(destinationObjectPtr);
```

Comment:
I am surprised this is not enforced with the compare and swap that happens in the setForwardedObject.

Reply:
Ok after some digging around I agree that this barrier is required on PPC HW. The compare and swap instruction following does not actually mean the data in the object will be visible but rather just that every thread would see the forwarding pointer. This storeSync forces the data in the newly copied object to be visible. Thanks @amicic